pkg/controllers/dnspolicy/dnspolicy_healthchecks.go

```diff
@@ -101,58 +104,46 @@ func (r *DNSPolicyReconciler) expectedProbesForGateway(ctx context.Context, gw c
-		dnsRecord := &v1alpha1.DNSRecord{}
-		if err := r.Client().Get(ctx, client.ObjectKey{Name: dnsRecordName(gw.Name, string(listener.Name)), Namespace: dnsPolicy.Namespace}, dnsRecord); err != nil {
-			if errors.IsNotFound(err) {
-				continue
-			} else {
-				return nil, err
-			}
+		matches := re.FindAllString(address.Value, -1)
+		if len(matches) != 1 {
+			log.V(3).Info("Found more than 1 ip, skipping")
+			continue
 		}
-		for _, endpoint := range dnsRecord.Spec.Endpoints {
 
-			log.V(1).Info("reconcileHealthChecks: found DNS Record endpoint", "endpoint", endpoint)
-			if endpoint.RecordType == string(v1alpha1.CNAMERecordType) {
-				// if the CNAME is a subdomain of this dnsrecord, skip it
-				if dns.IsSubDomain(dnsRecord.Name, endpoint.Targets[0]) {
-					continue
-				}
-			} else if endpoint.RecordType != string(v1alpha1.ARecordType) {
-				log.V(1).Info("reconcileHealthChecks: not an A record, skipping")
-				continue
-			}
-			for _, target := range endpoint.Targets {
-				port := dnsPolicy.Spec.HealthCheck.Port
-				if port == nil {
-					port = &defaultPort
-				}
-				log.V(1).Info("reconcileHealthChecks: adding health check for target", "target", target)
-				healthCheck := &v1alpha1.DNSHealthCheckProbe{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      fmt.Sprintf("%s-%s", target, dnsRecord.Name),
-						Namespace: gw.Namespace,
-						Labels:    commonDNSRecordLabels(client.ObjectKeyFromObject(gw), client.ObjectKeyFromObject(dnsPolicy)),
-					},
-					Spec: v1alpha1.DNSHealthCheckProbeSpec{
-						Port:                     *port,
-						Host:                     dnsRecord.Name,
-						Address:                  target,
-						Path:                     dnsPolicy.Spec.HealthCheck.Endpoint,
-						Protocol:                 *dnsPolicy.Spec.HealthCheck.Protocol,
-						Interval:                 metav1.Duration{Duration: 60 * time.Second},
-						AdditionalHeadersRef:     dnsPolicy.Spec.HealthCheck.AdditionalHeadersRef,
-						FailureThreshold:         dnsPolicy.Spec.HealthCheck.FailureThreshold,
-						ExpectedResponses:        dnsPolicy.Spec.HealthCheck.ExpectedResponses,
-						AllowInsecureCertificate: dnsPolicy.Spec.HealthCheck.AllowInsecureCertificates,
-					},
-				}
-				healthChecks = append(healthChecks, healthCheck)
-			}
+		log.V(1).Info("reconcileHealthChecks: adding health check for target", "target", address.Value)
+		healthCheck := &v1alpha1.DNSHealthCheckProbe{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", matches[0], dnsPolicy.Name),
+				Namespace: gw.Namespace,
+				Labels:    commonDNSRecordLabels(client.ObjectKeyFromObject(gw), client.ObjectKeyFromObject(dnsPolicy)),
+				Annotations: map[string]string{
+					AnnotationHealthCheckProbeEndpointName: dnsPolicy.Name,
+				},
+			},
+			Spec: v1alpha1.DNSHealthCheckProbeSpec{
+				Port:                     *port,
+				Host:                     string(*gw.Spec.Listeners[0].Hostname),
```

Comment:
should we not iterate over the listeners here, rather than hardcoding to only read the first one?

Reply:
Yes, sorry meant to make a note that this wasn't the final but just where I wanted to pitstop yesterday. 👍